File: pkg/api/verify_report.go
package api

import (
	"encoding/json"
	"fmt"
	"time"
)

// VerifyState is the terminal (or live) state of a verification report.
type VerifyState string

const (
	VerifyStateQueued    VerifyState = "queued"
	VerifyStateRunning   VerifyState = "running"
	VerifyStatePassed    VerifyState = "passed"
	VerifyStateFailed    VerifyState = "failed"
	VerifyStateErrored   VerifyState = "errored"
	VerifyStateWarned    VerifyState = "warned"
	VerifyStateSkipped   VerifyState = "skipped"
	VerifyStateCancelled VerifyState = "cancelled"
	VerifyStateTimedOut  VerifyState = "timed_out"
)

// AllVerifyStates lists every state in canonical order.
func AllVerifyStates() []VerifyState {
	return []VerifyState{
		VerifyStateQueued, VerifyStateRunning, VerifyStatePassed, VerifyStateFailed, VerifyStateErrored,
		VerifyStateWarned, VerifyStateSkipped, VerifyStateCancelled, VerifyStateTimedOut,
	}
}

// Validate rejects a state outside AllVerifyStates.
func (s VerifyState) Validate() error {
	for _, known := range AllVerifyStates() {
		if s == known {
			return nil
		}
	}
	return fmt.Errorf("invalid verify state %q", s)
}

// Report kinds produced by captain's own verifiers. A registry kind names the
// verifier family that produced a report; hosts add their own (e.g. "fixture").
const (
	VerifyKindCmd     = "cmd"
	VerifyKindPrompt  = "prompt"
	VerifyKindFunc    = "func"
	VerifyKindFixture = "fixture"
	// VerifyKindRound is what MergeReports stamps on a round whose reports do not
	// share one kind: the merged document is a whole verification round rather
	// than the output of any one verifier family.
	VerifyKindRound = "round"
)

// VerifySummary counts the leaves of a report's test tree. It is the wire twin
// of clicky-ui's TestSummary — a node carrying one is summarised from it rather
// than from its children (see VerifyNode.Summary) — and SummarizeNodes tallies
// leaves exactly as that package's countsFromLeaf does, timed-out bucket
// included.
type VerifySummary struct {
	Total   int `json:"total"`
	Passed  int `json:"passed"`
	Failed  int `json:"failed"`
	Warned  int `json:"warned"`
	Skipped int `json:"skipped"`
	Pending int `json:"pending"`
	Running int `json:"running"`
	// TimedOut is its own bucket rather than a flavour of failed, and is spelled
	// `timedout` because that is the key clicky-ui's StatusCounts reads.
	TimedOut int `json:"timedout"`
}

// VerifyNodeProgress is a running node's in-flight position.
type VerifyNodeProgress struct {
	Phase string `json:"phase,omitempty"`
	Done  int    `json:"done"`
	Total int    `json:"total"`
}

// VerifyNodeContext mirrors clicky-ui's FixtureContext: what a command-shaped
// node ran and, for a CEL expectation, what it evaluated.
type VerifyNodeContext struct {
	Command       string         `json:"command,omitempty"`
	ExitCode      int            `json:"exit_code"`
	Cwd           string         `json:"cwd,omitempty"`
	CELExpression string         `json:"cel_expression,omitempty"`
	CELVars       map[string]any `json:"cel_vars,omitempty"`
	Expected      any            `json:"expected,omitempty"`
	Actual        any            `json:"actual,omitempty"`
}

// VerifyNode is one node of a verification tree. Its JSON is the snake_case
// wire shape of clicky-ui's Test so a TestRunner renders it unchanged; a node
// with Children is a group and carries no verdict of its own.
type VerifyNode struct {
	Name      string              `json:"name"`
	Framework string              `json:"framework,omitempty"`
	TaskID    string              `json:"task_id,omitempty"`
	File      string              `json:"file,omitempty"`
	Line      int                 `json:"line,omitempty"`
	Message   string              `json:"message,omitempty"`
	Command   string              `json:"command,omitempty"`
	WorkDir   string              `json:"work_dir,omitempty"`
	Stdout    string              `json:"stdout,omitempty"`
	Stderr    string              `json:"stderr,omitempty"`
	Duration  time.Duration       `json:"duration,omitempty"`
	Passed    bool                `json:"passed,omitempty"`
	Failed    bool                `json:"failed,omitempty"`
	Warned    bool                `json:"warned,omitempty"`
	Skipped   bool                `json:"skipped,omitempty"`
	Pending   bool                `json:"pending,omitempty"`
	Running   bool                `json:"running,omitempty"`
	TimedOut  bool                `json:"timed_out,omitempty"`
	Progress  *VerifyNodeProgress `json:"progress,omitempty"`
	Context   *VerifyNodeContext  `json:"context,omitempty"`
	Detail    json.RawMessage     `json:"detail,omitempty"`
	// Summary is this node's own tally, and it wins over its children — exactly
	// as clicky-ui's sum() reads `t.summary` before it recurses. It is what lets
	// a producer ship the counts of a suite whose rows it elided (a merged round
	// nests one group per report; a 40-test fixture may send its totals and only
	// the failures), where recursing would report that suite as empty.
	Summary  *VerifySummary `json:"summary,omitempty"`
	Children []VerifyNode   `json:"children,omitempty"`
}

// VerifyChecklistItem is one acceptance-criteria verdict. Passed is nil while
// the item has not been judged.
type VerifyChecklistItem struct {
	Item    string `json:"item"`
	Passed  *bool  `json:"passed"`
	Message string `json:"message,omitempty"`
}

// VerifyReport is a verifier's typed judgement: the verdict, the tree of what
// ran, and the acceptance-criteria checklist. It is what a Verify hook returns,
// what captain persists per iteration, and what the webapp renders.
type VerifyReport struct {
	Kind     string `json:"kind"`
	Name     string `json:"name,omitempty"`
	Ran      bool   `json:"ran"`
	Passed   bool   `json:"passed"`
	Reason   string `json:"reason,omitempty"`
	Feedback string `json:"feedback,omitempty"`
	// Iteration is the 1-based loop turn this report judged ("turn 1 of 3"), the
	// same numbering captain_prompt_run_iterations is keyed on. It is always on
	// the wire: with omitempty, an unstamped report and the first turn's report
	// arrived as the same document, and the store cannot tell them apart.
	Iteration  int                   `json:"iteration"`
	Summary    VerifySummary         `json:"summary"`
	Tests      []VerifyNode          `json:"tests,omitempty"`
	Checklist  []VerifyChecklistItem `json:"checklist,omitempty"`
	State      VerifyState           `json:"state"`
	StartedAt  *time.Time            `json:"started_at,omitempty"`
	FinishedAt *time.Time            `json:"finished_at,omitempty"`
	Duration   time.Duration         `json:"duration,omitempty"`
}

// SummarizeNodes counts the leaves of a tree; a node with children is a group,
// not a result. It mirrors clicky-ui's sum() exactly, so the counters captain
// persists and the ones the webapp recomputes never disagree:
//
//   - a node carrying its own Summary is counted from it and never recursed
//     into, so an elided child list still contributes the whole suite;
//   - a timed-out leaf counts only in TimedOut, never in Failed;
//   - a leaf carrying no status flag at all is not counted, Total included — it
//     is a placeholder row, not a queued test;
//   - otherwise the leaf's own flags are counted, one bucket each.
func SummarizeNodes(nodes []VerifyNode) VerifySummary {
	var s VerifySummary
	for i := range nodes {
		n := &nodes[i]
		switch {
		case n.Summary != nil:
			s = AddSummaries(s, *n.Summary)
		case len(n.Children) > 0:
			s = AddSummaries(s, SummarizeNodes(n.Children))
		default:
			s = AddSummaries(s, summarizeLeaf(n))
		}
	}
	return s
}

// summarizeLeaf is clicky-ui's countsFromLeaf: timed out wins outright, and a
// flagless leaf contributes nothing.
func summarizeLeaf(n *VerifyNode) VerifySummary {
	if n.TimedOut {
		return VerifySummary{Total: 1, TimedOut: 1}
	}
	var s VerifySummary
	if n.Passed {
		s.Passed = 1
	}
	if n.Failed {
		s.Failed = 1
	}
	if n.Warned {
		s.Warned = 1
	}
	if n.Skipped {
		s.Skipped = 1
	}
	if n.Pending {
		s.Pending = 1
	}
	if n.Running {
		s.Running = 1
	}
	if s != (VerifySummary{}) {
		s.Total = 1
	}
	return s
}

// AddSummaries totals two summaries, bucket by bucket, so a caller rolling
// several reports into one verdict never re-lists the fields (and never forgets
// the one added last).
func AddSummaries(into, from VerifySummary) VerifySummary {
	into.Total += from.Total
	into.Passed += from.Passed
	into.Failed += from.Failed
	into.Warned += from.Warned
	into.Skipped += from.Skipped
	into.Pending += from.Pending
	into.Running += from.Running
	into.TimedOut += from.TimedOut
	return into
}

// StateForNode derives a state from one node — the one-node case of
// StateForReport, and defined as it rather than as a second hand-ordered switch.
// Two switches drifted on the leaves that carry contradictory flags (a node both
// skipped and running, both pending and skipped): NewNodeReport stamped the
// state one of them chose and Validate then rejected the report against the
// other. There is one precedence, and it lives in StateForReport.
func StateForNode(n VerifyNode) VerifyState {
	return StateForReport([]VerifyNode{n})
}

// StateForReport derives a report's state from its whole tree, in the order a
// reader cares about: an outright failure first, then a check that never
// finished, then a soft failure, then anything still moving, then the states a
// finished-and-uneventful tree can be in. An empty tree has not started, so it
// is queued.
//
// It is the one definition of that precedence: StateForNode is this function
// applied to a single node, so the state NewNodeReport stamps and the one
// Validate checks can never disagree.
func StateForReport(tests []VerifyNode) VerifyState {
	s := SummarizeNodes(tests)
	switch {
	case s.Failed > 0:
		return VerifyStateFailed
	case s.TimedOut > 0:
		return VerifyStateTimedOut
	case s.Warned > 0:
		return VerifyStateWarned
	case s.Running > 0:
		return VerifyStateRunning
	case s.Pending > 0:
		return VerifyStateQueued
	case s.Skipped > 0:
		return VerifyStateSkipped
	case s.Passed > 0:
		return VerifyStatePassed
	default:
		return VerifyStateQueued
	}
}

// NewNodeReport builds a report from a single leaf: the verdict is the leaf's
// state, and only a passed leaf passes.
func NewNodeReport(kind, name string, node VerifyNode) VerifyReport {
	state := StateForNode(node)
	return VerifyReport{
		Kind:     kind,
		Name:     name,
		Ran:      state != VerifyStateQueued,
		Passed:   state == VerifyStatePassed,
		Reason:   node.Message,
		Summary:  SummarizeNodes([]VerifyNode{node}),
		Tests:    []VerifyNode{node},
		State:    state,
		Duration: node.Duration,
	}
}

// HostStamped reports whether the state is one the producing host asserts
// rather than one the tree implies. A runner that could not schedule its nodes
// (errored) or a run stopped mid-check (cancelled) leaves queued leaves behind
// that say nothing about why; no node flag maps to either, so Validate takes
// them as stamped. Neither can pass.
func (s VerifyState) HostStamped() bool {
	return s == VerifyStateErrored || s == VerifyStateCancelled
}

// Validate checks the report's internal consistency: only a passed report
// passes, a report that never ran is queued, running, errored or cancelled,
// the summary matches the leaves of its tree, and — unless the host stamped a
// terminal state of its own — the state is the one that tree justifies.
//
// The state check is what stops a red run reading as green downstream: the
// webapp colours its badge from State while the panel lists the tree, and a CEL
// predicate written against the state would pass on a report whose tests failed.
//
// A live snapshot is a valid report: a ProgressVerifier publishes one that is
// already running before it has run, so `running` joins `queued` as a state a
// not-yet-finished report is allowed to be in.
func (r VerifyReport) Validate() error {
	if r.Kind == "" {
		return fmt.Errorf("verify report: kind is required")
	}
	if err := r.State.Validate(); err != nil {
		return fmt.Errorf("verify report %q: %w", r.Name, err)
	}
	if r.Passed && r.State != VerifyStatePassed {
		return fmt.Errorf("verify report %q: passed=true with state %q", r.Name, r.State)
	}
	if want := SummarizeNodes(r.Tests); want != r.Summary {
		return fmt.Errorf("verify report %q: summary %+v does not match its %d leaf node(s) %+v", r.Name, r.Summary, want.Total, want)
	}
	if r.State.HostStamped() {
		return nil
	}
	if !r.Ran && r.State != VerifyStateQueued && r.State != VerifyStateRunning {
		return fmt.Errorf("verify report %q: ran=false with state %q; a report that did not run is queued, running, errored or cancelled", r.Name, r.State)
	}
	if len(r.Tests) > 0 {
		if want := StateForReport(r.Tests); want != r.State {
			return fmt.Errorf("verify report %q: state %q but its tests are %q", r.Name, r.State, want)
		}
	}
	return nil
}

// CELVars renders the report as the plain map a host binds to the CEL variable
// `verify` (wire field names, numbers as float64) so predicates such as
// `verify.summary.failed > 0` and `verify.checklist.all(i, i.passed)` read the
// same shape the webapp does.
//
// A report a verifier built with an unencodable detail is an error rather than
// a panic: the caller is a host evaluating a predicate, and it can report a
// broken report far better than a stack unwinding through its evaluator can.
func (r VerifyReport) CELVars() (map[string]any, error) {
	raw, err := json.Marshal(r)
	if err != nil {
		return nil, fmt.Errorf("verify report %q: marshal: %w", r.Name, err)
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("verify report %q: unmarshal: %w", r.Name, err)
	}
	if out["checklist"] == nil {
		out["checklist"] = []any{}
	}
	if out["tests"] == nil {
		out["tests"] = []any{}
	}
	return out, nil
}

File: pkg/cli/webapp/src/types/verifyReport.ts
import type {
  VerifyChecklistItem,
  VerifyNode,
  VerifyNodeContext,
  VerifyNodeProgress,
  VerifyReport,
  VerifyState,
  VerifySummary,
} from "@flanksource/clicky-ui/data";

export type { VerifyReport } from "@flanksource/clicky-ui/data";

/** Canonical order mirrors Go's `AllVerifyStates()`. */
export const VERIFY_STATES = [
  "queued",
  "running",
  "passed",
  "failed",
  "errored",
  "warned",
  "skipped",
  "cancelled",
  "timed_out",
] as const;

/** The `verify` SSE event payload: the newest report, and whether it is the
 * verdict (`done: true`) or a still-running snapshot (`done: false`). */
export interface VerifyFrame {
  report: VerifyReport | null;
  done: boolean;
}

/**
 * parseVerifyFrame validates an already-JSON-decoded value against the
 * VerifyFrame wire shape and throws a descriptive error on any drift — a
 * missing field, a wrong type, or a state string the Go side no longer emits.
 * There is no silent default: a malformed frame must fail loudly rather than
 * render a stale or blank verification panel.
 */
export function parseVerifyFrame(data: unknown): VerifyFrame {
  const frame = requireRecord(data, "verify frame");
  const done = requireBoolean(frame.done, 'verify frame: "done"');
  const report =
    frame.report === null || frame.report === undefined
      ? null
      : parseVerifyReport(frame.report);
  return { report, done };
}

function parseVerifyReport(value: unknown): VerifyReport {
  const r = requireRecord(value, "verify report");
  const state = requireString(r.state, 'verify report: "state"');
  if (!isVerifyState(state)) {
    throw new Error(
      `verify report: "state" must be one of ${VERIFY_STATES.join(", ")}, got ${JSON.stringify(state)}`,
    );
  }
  return {
    kind: requireString(r.kind, 'verify report: "kind"'),
    name: optionalString(r.name, 'verify report: "name"'),
    ran: requireBoolean(r.ran, 'verify report: "ran"'),
    passed: requireBoolean(r.passed, 'verify report: "passed"'),
    reason: optionalString(r.reason, 'verify report: "reason"'),
    feedback: optionalString(r.feedback, 'verify report: "feedback"'),
    iteration: requireNumber(r.iteration, 'verify report: "iteration"'),
    summary: parseVerifySummary(r.summary),
    tests: optionalArray(r.tests, 'verify report: "tests"', parseVerifyNode),
    checklist: optionalArray(
      r.checklist,
      'verify report: "checklist"',
      parseChecklistItem,
    ),
    state,
    started_at: optionalString(r.started_at, 'verify report: "started_at"'),
    finished_at: optionalString(r.finished_at, 'verify report: "finished_at"'),
    duration: optionalNumber(r.duration, 'verify report: "duration"'),
  };
}

function parseVerifySummary(value: unknown): VerifySummary {
  const s = requireRecord(value, "verify report summary");
  return {
    total: requireNumber(s.total, 'verify report summary: "total"'),
    passed: requireNumber(s.passed, 'verify report summary: "passed"'),
    failed: requireNumber(s.failed, 'verify report summary: "failed"'),
    warned: requireNumber(s.warned, 'verify report summary: "warned"'),
    skipped: requireNumber(s.skipped, 'verify report summary: "skipped"'),
    pending: requireNumber(s.pending, 'verify report summary: "pending"'),
    running: requireNumber(s.running, 'verify report summary: "running"'),
    timedout: requireNumber(s.timedout, 'verify report summary: "timedout"'),
  };
}

function parseVerifyNode(value: unknown): VerifyNode {
  const n = requireRecord(value, "verify test node");
  return {
    name: requireString(n.name, 'verify test node: "name"'),
    framework: optionalString(n.framework, 'verify test node: "framework"'),
    task_id: optionalString(n.task_id, 'verify test node: "task_id"'),
    file: optionalString(n.file, 'verify test node: "file"'),
    line: optionalNumber(n.line, 'verify test node: "line"'),
    message: optionalString(n.message, 'verify test node: "message"'),
    command: optionalString(n.command, 'verify test node: "command"'),
    work_dir: optionalString(n.work_dir, 'verify test node: "work_dir"'),
    stdout: optionalString(n.stdout, 'verify test node: "stdout"'),
    stderr: optionalString(n.stderr, 'verify test node: "stderr"'),
    duration: optionalNumber(n.duration, 'verify test node: "duration"'),
    passed: optionalBoolean(n.passed, 'verify test node: "passed"'),
    failed: optionalBoolean(n.failed, 'verify test node: "failed"'),
    warned: optionalBoolean(n.warned, 'verify test node: "warned"'),
    skipped: optionalBoolean(n.skipped, 'verify test node: "skipped"'),
    pending: optionalBoolean(n.pending, 'verify test node: "pending"'),
    running: optionalBoolean(n.running, 'verify test node: "running"'),
    timed_out: optionalBoolean(n.timed_out, 'verify test node: "timed_out"'),
    progress: n.progress === undefined ? undefined : parseProgress(n.progress),
    context: n.context === undefined ? undefined : parseContext(n.context),
    summary: n.summary === undefined ? undefined : parseVerifySummary(n.summary),
    detail: n.detail,
    children: optionalArray(
      n.children,
      'verify test node: "children"',
      parseVerifyNode,
    ),
  };
}

function parseProgress(value: unknown): VerifyNodeProgress {
  const p = requireRecord(value, "verify test node progress");
  return {
    phase: optionalString(p.phase, 'verify test node progress: "phase"'),
    done: requireNumber(p.done, 'verify test node progress: "done"'),
    total: requireNumber(p.total, 'verify test node progress: "total"'),
  };
}

function parseContext(value: unknown): VerifyNodeContext {
  const c = requireRecord(value, "verify test node context");
  return {
    command: optionalString(c.command, 'verify test node context: "command"'),
    exit_code: requireNumber(
      c.exit_code,
      'verify test node context: "exit_code"',
    ),
    cwd: optionalString(c.cwd, 'verify test node context: "cwd"'),
    cel_expression: optionalString(
      c.cel_expression,
      'verify test node context: "cel_expression"',
    ),
    cel_vars:
      c.cel_vars === undefined
        ? undefined
        : requireRecord(c.cel_vars, 'verify test node context: "cel_vars"'),
    expected: c.expected,
    actual: c.actual,
  };
}

function parseChecklistItem(value: unknown): VerifyChecklistItem {
  const c = requireRecord(value, "verify report checklist item");
  const passed = c.passed;
  if (passed !== null && typeof passed !== "boolean") {
    throw new Error(
      `verify report checklist item: "passed" must be a boolean or null, got ${typeOf(passed)}`,
    );
  }
  return {
    item: requireString(c.item, 'verify report checklist item: "item"'),
    passed,
    message: optionalString(
      c.message,
      'verify report checklist item: "message"',
    ),
  };
}

function isVerifyState(value: string): value is VerifyState {
  return (VERIFY_STATES as readonly string[]).includes(value);
}

function typeOf(value: unknown): string {
  if (value === null) return "null";
  if (Array.isArray(value)) return "array";
  return typeof value;
}

function requireRecord(
  value: unknown,
  label: string,
): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${label}: expected an object, got ${typeOf(value)}`);
  }
  return value as Record<string, unknown>;
}

function requireString(value: unknown, label: string): string {
  if (typeof value !== "string") {
    throw new Error(`${label} must be a string, got ${typeOf(value)}`);
  }
  return value;
}

function optionalString(value: unknown, label: string): string | undefined {
  if (value === undefined) return undefined;
  return requireString(value, label);
}

function requireNumber(value: unknown, label: string): number {
  if (typeof value !== "number") {
    throw new Error(`${label} must be a number, got ${typeOf(value)}`);
  }
  return value;
}

function optionalNumber(value: unknown, label: string): number | undefined {
  if (value === undefined) return undefined;
  return requireNumber(value, label);
}

function requireBoolean(value: unknown, label: string): boolean {
  if (typeof value !== "boolean") {
    throw new Error(`${label} must be a boolean, got ${typeOf(value)}`);
  }
  return value;
}

function optionalBoolean(value: unknown, label: string): boolean | undefined {
  if (value === undefined) return undefined;
  return requireBoolean(value, label);
}

function optionalArray<T>(
  value: unknown,
  label: string,
  parseItem: (item: unknown) => T,
): T[] | undefined {
  if (value === undefined) return undefined;
  if (!Array.isArray(value)) {
    throw new Error(`${label} must be an array, got ${typeOf(value)}`);
  }
  return value.map(parseItem);
}
